storage-proofs/src/hybrid_merkle.rs

```diff
@@ -0,0 +1,1445 @@
+where
+    A: Domain,
+    B: Domain,
+{
+    let fr: Fr = elem.into();
+    B::from(fr)
+}
+
+/// Copies the functionality used by `merkle_light::merkle::MerkleTree` to hash a node (non-leaf) in
+/// the tree.
+pub fn hash_node<H>(left: H::Domain, right: H::Domain, layer: usize) -> H::Domain
+where
+    H: Hasher,
+{
+    let mut alg = H::Function::default();
+    alg.node(left, right, layer)
+}
+
+/// Stores the value for a node in a Hybrid Merkle Tree . If an node is in one of the beta hasher's
+/// trees, then the field `beta` will be `Some`, likewise if the element is found in the alpha
+/// hasher's tree, then `alpha` will be set to `Some`. If an element is the root of a beta tree/leaf
+/// of the alpha tree, then both `beta` and `alpha` will be set.
+///
+/// This type is returned by `HybridMerkleTree::read_at()`. `HybridMerkleTree` does not use
+/// `HybridNode` internally to store elements.
+#[derive(Debug, Default)]
+pub struct HybridNode<AH, BH>
+where
+    AH: Hasher,
+    BH: Hasher,
+{
+    pub beta: Option<BH::Domain>,
+    pub alpha: Option<AH::Domain>,
+}
+
+/// Represents the positional information associated with a node in a `HybridMerkleTree`.
+///
+/// A `HybridMerkleTree` is composed of smaller subtrees (many beta trees and one alpha tree). These
+/// subtrees are abstracted away from the user. The user views a Hybrid Merkle Tree the same as they
+/// would a regular Merkle Tree. A user references a node in a Hybrid Merkle Tree using a node's
+/// "abstracted" node index, however nodes are not actually stored at that index (because Hybrid
+/// Merkle Trees are composed of many smaller subtrees). Therefore, some node index
+/// arithmetic/conversion must happen to convert the abstracted user-facing node index to a node
+/// index in a subtree. The output of some of these arithmetic steps are utilized in multiple
+/// `HybridMerkleTree` methods, we compute these values once and use this structure to pass the
+/// values around.
+#[derive(Debug)]
+struct NodeInfo {
+    /// This node's abstracted node index in a Hybrid Merkle Tree.
+    node_index: usize,
+
+    /// The Hybrid Merkle Tree layer that contains the node (layer 0 is the base/leaf layer, layer 1
+    /// is the layer above it, etc...).
+    layer: usize,
+
+    /// Is this node part of a beta subtree.
+    is_in_beta_tree: bool,
+
+    /// Is this node part of the alpha subtree.
+    is_in_alpha_tree: bool,
+
+    /// If this node is part of a beta tree then this field gives the index in
+    /// `HybridMerkleTree.beta_trees` for the beta tree containing this node.
+    beta_tree_index: Option<usize>,
+
+    /// If this node is part of a beta tree, then this field gives the node's node index in the beta
+    /// tree.
+    beta_tree_node_index: Option<usize>,
+
+    /// If this node is part of the alpha tree, then this field gives the node's node index in the
+    /// alpha tree.
+    alpha_tree_node_index: Option<usize>,
```

Comment:
You would still need both a `alpha_subtree: Option<SubtreeInfo>` and a `beta_subtree: Option<SubtreeInfo>` because a node can be in both an alpha and beta subtree

Reply:
but that could be explicitly one of the `SubtreeInfo` cases

Reply:
So like?

```rust
enum SubtreeInfo {
    Alpha {
        node_index: usize,
    },
    Beta {
        tree_index: usize,
        node_index: usize,
    },
    Both {
        beta_tree_index: usize,
        beta_node_index: usize,
        alpha_node_index: usize,
    },
}
```

That is more code, but it probably is cleaner to have the three states written explicitly.